hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java

```diff
@@ -0,0 +1,249 @@
+
+import static com.hazelcast.spi.Operation.CALL_ID_LOCAL_SKIPPED;
+import static com.hazelcast.spi.OperationAccessor.setCallTimeout;
+
+/**
+ * Responsible for generating callIds.
+ * <p/>
+ * It is very important that for each {@link #next(Invocation)}, there is a matching {@link #complete(Invocation)}.
+ * If this doesn't match, than either the number of concurrent invocations can grow or shrink over time. This can lead
+ * to OOME or deadlock.
+ * <p/>
+ * The CallIdSequence provides back pressure if enabled and there are too many concurrent invocations. When this happens,
+ * an exponential backoff policy is applied.
+ * <p/>
+ * Currently a single CallIdSequence is used for all partitions; so there is contention. Also one partition can cause problems
+ * in other partition if a lot of invocations are created for that partition. Then other partitions can't make as many invocations
+ * because a single callIdSequence is being used.
+ * <p/>
+ * In the future we could add a CallIdSequence per partition or using some 'concurrency level' and do a mod based on the
+ * partition-id. The advantage is that you reduce contention and improved isolation, at the expensive of:
+ * <ol>
+ *     <li>increased complexity</li>
+ *     <li>not always being able to fully utilize the number of invocations.</li>
+ * </ol>
+ *
+ */
+public abstract class CallIdSequence {
+
+    // just for testing purposes.
+    abstract long getLastCallId();
+
+    /**
+     * Returns the maximum concurrent invocations supported. Integer.MAX_VALUE means there is no max.
+     *
+     * @return the maximum concurrent invocation.
+     */
+    public abstract int getMaxConcurrentInvocations();
+
+    /**
+     * Creates the next call-id.
+     * <p/>
+     * {@link com.hazelcast.spi.Operation#CALL_ID_LOCAL_SKIPPED} is returned if local registration in
+     * the InvocationRegistry can be skipped.
+     *
+     * @param invocation the Invocation to create a call-id for.
+     * @return the generated callId.
+     * @throws AssertionError if invocation.op.callId not equals 0
+     */
+    public abstract long next(Invocation invocation);
+
+    public abstract void complete(Invocation invocation);
+
+    /**
+     * Not every call needs to be registered. A call that is local and has no backups, doesn't need to be registered
+     * since there will not be a remote machine sending a response back to the invocation.
+     *
+     * @param invocation
+     * @return true if registration is required, false otherwise.
+     */
+    protected boolean skipRegistration(Invocation invocation) {
+        if (invocation.remote) {
+            return false;
+        }
+
+        if (invocation.op instanceof BackupAwareOperation) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public static final class CallIdSequenceWithoutBackpressure extends CallIdSequence {
+
+        private static final AtomicLongFieldUpdater<CallIdSequenceWithoutBackpressure> HEAD
```

Comment:
Strange.. I thought I left a comment here.

There is one less memory indirection. I don't want to make the previous situation slower than the current one. So instead of having an atomiclong with a volatile field, you have a CallSequenceIdWithoutBackPressure with volatile field.